### ISSUE TITLE
lua 5.2, __pairs and __ipairs metatable methods polyfill

### ISSUE THREAD

**Issue by wsor4035** [minetest/minetest#15133]
### Problem

ability to have meta access to when a table is looped over with pair/ipairs

additional context: https://github.com/minetest/minetest/issues/14821#issuecomment-2218349444

### Solutions

insert https://stackoverflow.com/a/77354254 in builtin somewhere and make sure it doesnt have any performance implications

### Alternatives

dont? let mods polyfill this themselves?

### Additional context

(side note, marking this as supported by core dev for lars via discord)

**Comment by appgurueu** [comment]
Notably LuaJIT already has this 5.2 extension (see https://luajit.org/extensions.html), so this polyfill must not be enabled there and then there is no performance concern on LuaJIT at least.

Some rudimentary tests to make sure the 5.1 polyfill isn't wrecking 5.1 `pairs` / `ipairs` performance for small tables should be done, but that is less important than LuaJIT performance.

Some notes (looking at the linked SO snippet):

* `getmetatable` should probably be localized.
* Could write this as a "ternary" via `and`-`or` (mutually exclusive with the next one).
* Could also save a function call by replacing `pairs(t)` with `next, t`. Similar for `ipairs`.

---

Offtopic for this issue, however two further possibly interesting things to explore when it comes to strictness:

* It is possible to randomize `pairs` to prevent people from relying on iteration order by randomly skipping the first key-value pair.
* It is possible to have `ipairs` throw a warning when it didn't iterate all keys by relying on implementation details of `next`.

Though these might be better left to a mod. They would need to be optional and off by default anyways.

**Comment by red-001** [comment]
Won't it be easier to backport this into Lua 5.1 proper and not rely on that linked hack? Minetest already doesn't use actual PUC Lua, might as well make this change in C/C++.

**Comment by wsor4035** [comment]
if going to go that route, long term might as well match the other luajit 5.2 baclports/differences at https://luajit.org/extensions.html

**Comment by red-001** [comment]
that seems like a good idea, there's probably a decent bit that can be backported in general from the Lua internals. I'm going to look into the pairs thing and the `load` changes for an initial backport PR.

**Comment by appgurueu** [comment]
> Won't it be easier to backport this into Lua 5.1 proper and not rely on that linked hack? Minetest already doesn't use actual PUC Lua, might as well make this change in C/C++.

Probably a preferable option, including performance concerns. Only minor "downside" I see would be that going down this route makes #15102 harder. In principle I'm open to us maintaining our own Lua 5.1 with some 5.2 sprinkled in fork. 

We *could* even consider going for full 5.2 support (most of which has already been available on LuaJIT) if we provide a few things for backwards compatibility, like a `setfenv` polyfill.

**Comment by red-001** [comment]
I suspect #15102 was just not going to happen (in a meaningful way) anyways, Lua itself does not accept patches so what good for maintenance does getting some other Lua users on board really do?

And a lot of Lua embeders whatever closed or open source end up modifying the language so there isn't much of a chance for an official "community edition" of Lua 5.1.

**Comment by sfan5** [comment]
> We _could_ even consider going for full 5.2 support (most of which has already been available on LuaJIT)

#1277

**Comment by Desour** [comment]
>Notably LuaJIT already has this 5.2 extension (see https://luajit.org/extensions.html), so this polyfill must not be enabled there and then there is no performance concern on LuaJIT at least.

It's listed under
> Other features are only enabled, if LuaJIT is built with `-DLUAJIT_ENABLE_LUA52COMPAT`:

**Comment by red-001** [comment]
Could we maybe ask LuaJIT upstream for a feature flag to enable just the pairs functionality? Or would we be happy enabling all those new features? It seems like a fairly manageable backcompat break.

**Comment by wsor4035** [comment]
Pretty sure there enabled by default, hence the issue with mods using gito, etc and causing issues

**Comment by sfan5** [comment]
> Or would we be happy enabling all those new features?

NB: we don't decide which flags luajit is built with on Linux. distributions do.

**Comment by okias** [comment]
Distros usually (just checked Debian and Alpine), use [openResty LUAJIT](https://openresty.org/en/luajit.html).

Just snippet from Alpine:
```
        # These flags are recommended by OpenResty.
        local xcflags="-DLUAJIT_ENABLE_LUA52COMPAT -DLUAJIT_NUMMODE=2"
```

Debian doesn't [use the flag](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=781728), but I believe now there should be more pressure to add it, so I see that as a viable option.

**Comment by sfan5** [comment]
Thanks for checking.
I thought it was enabled by default actually but it isn't:
```
/tmp/luajit $ rg DLUAJIT_ENABLE_LUA52COMPAT
doc/extensions.html
369:<tt>-DLUAJIT_ENABLE_LUA52COMPAT</tt>:

src/Makefile
101:#XCFLAGS+= -DLUAJIT_ENABLE_LUA52COMPAT
```

For this issue that means adding a polyfill to our own Lua source would actually *worsen* the compatibility.

**Comment by sfan5** [comment]
@wsor4035 @appgurueu which concrete cases do you actually see for using this? are they important enough?

It really looks like implementing this would be maneuvering us into a compatibility mess.

**Comment by okias** [comment]
> It really looks like implementing this would be maneuvering us into a compatibility mess.

You can always ship luajit with `-DLUAJIT_ENABLE_LUA52COMPAT` and distros who has the package can avoid custom dependency. IMHO that doesn't sounds that bad.

If I get it right that would mean 3 options:
 - bundled luajit
 - distribution luajit w/`-DLUAJIT_ENABLE_LUA52COMPAT`
 - distribution lua 5.2 for luajit unsupported archs

Right now, we have custom patched

**Comment by appgurueu** [comment]
> @wsor4035 @appgurueu which concrete cases do you actually see for using this?

Essentially whenever you have some kind of "proxy" object that wants to provide an iteration interface that is compatible with tables.

> are they important enough?

Not necessarily, but they would allow proxies to act like tables much better.

> It really looks like implementing this would be maneuvering us into a compatibility mess.

The alternative is, as was brought up initially, to implement this in Lua like this:

```lua
local __pairs_called = false
pairs(setmetatable({}, {__pairs = function() __pairs_called = true end}))
if not __pairs_called then
    local raw_pairs = pairs
    local getmetatable = getmetatable
    function pairs(t)
        local mt = getmetatable(t)
        if mt and mt.__pairs then return mt.__pairs(t) end
        return raw_pairs(t)
    end
end
```

and the same for `__ipairs`, which would raise no compatibility issues. I haven't benchmarked this yet though, but note that if `__pairs` is available (because LuaJIT was compiled with Lua 5.2 compat), it will just rely on that and there will be no slowdown.

**Comment by appgurueu** [comment]
Prompted by https://content.luanti.org/packages/TPH/metatable_metamethods/ and for anyone who might take a look at this in the future, I need to correct a few details:

* My snippet above has gotten at least two (hopefully usually minor) details wrong:
  1. `getmetatable` will not return a metatable hidden via `__metatable`, you would need to use `debug.getmetatable` for that;
  2. It should not just index the metatable: It should `rawget` instead.
* There are at least two major issues with trying to implement this satisfactorily in a mod:
  1. If other mods have loaded before you, they may have localized `pairs`, in which case there is nothing you can do.
  2. With mod security, you only have access to `getmetatable`, not `debug.getmetatable`.

---

Furthermore, a note about `ipairs`. This is an excerpt from the Lua 5.3 changelog:

> The [ipairs](https://www.lua.org/manual/5.3/manual.html#pdf-ipairs) iterator now respects metamethods and its `__ipairs` metamethod has been deprecated".

Indeed `ipairs` simply respecting `__index` is a pretty clean solution (note that `__len` is not needed, `t[#t + 1]` should just be `nil`).
